tests/test_common.py

```diff
@@ -123,6 +124,31 @@ def test_atomic_directory_empty_workdir_finalized():
             ), "When the target_dir exists no work_dir should be created."
 
 
+def test_atomic_directory_deadlock():
+    # type: () -> None
+
+    def mock_lockf(fd, cmd):  # type: (int, int) -> None
```

Comment:
Instead of mocking we could try recreating the exact series of events, somewhat serially, which leads to this behavior but I'm not confident the code would be very helpful (likely involving some gnarly signaling between threads/processes) and I'm also not confident would reproduce this case on every OS.